### PR TITLE
Movie gen

### PR DESCRIPTION
**Pull request by hadipash**

### CODE REVIEW

File: examples/moviegen/moviegen/dataset/dataset.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+import numpy as np
+from tqdm import tqdm
+
+from mindspore.dataset.transforms import Compose
+
+from mindone.data.video_reader import VideoReader
+
+from .transforms import ResizeCrop
+
+# FIXME: remove in future when mindone is ready for install
```
Comment (zhtmike) [inline examples/moviegen/moviegen/dataset/dataset.py:20]: no need to add path here, unless it is a entry file with `__name__ = "__main__"`

File: examples/moviegen/moviegen/dataset/dataset.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+sys.path.append(os.path.join(os.path.dirname(__file__), "../../../.."))
+from mindone.data import BaseDataset
+
+_logger = logging.getLogger(__name__)
+
+
+IMAGE_EXT = (".jpg", ".jpeg", ".png", ".gif", ".webp")
+
+
+def create_infer_transforms(target_size: Tuple[int, int], interpolation=cv2.INTER_LINEAR):
```
Comment (zhtmike) [inline examples/moviegen/moviegen/dataset/dataset.py:30]: `create_infer_transforms` seems has no usage now, perhaps drop it?

File: examples/moviegen/moviegen/models/llama/network.py [hunk 3]
```diff
@@ -459,10 +461,7 @@ def unpatchify(self, hidden_states: Tensor, t: int, h: int, w: int) -> Tensor:
         return output
 
     def construct(
-        self,
-        latent_embedding: Tensor,
-        timestep: Tensor,
-        text_embedding: Tensor,
+        self, latent_embedding: Tensor, timestep: Tensor, ul2_emb: Tensor, metaclip_emb: Tensor, byt5_emb: Tensor
```
Comment (zhtmike) [inline examples/moviegen/moviegen/models/llama/network.py:464]: in docstring, may need add shape info for the inputs for easier development

File: examples/moviegen/moviegen/pipelines/train_pipeline.py [hunk 4]
```diff
@@ -42,16 +41,24 @@ def __init__(
...
+            text_emb = ops.stop_gradient(self.text_encoder(text_tokens))
         return text_emb
 
     def get_latents(self, video_tokens: Tensor) -> Tensor:
         if self.video_emb_cached:
             return video_tokens
-        video_emb = ops.stop_gradient(self.vae.encode(video_tokens) * self.scale_factor)
+        with no_grad():
+            # (b c f h w) shape is expected. FIXME: remove this redundancy
+            video_tokens = ops.transpose(video_tokens, (0, 2, 1, 3, 4))
```
Comment (zhtmike) [inline examples/moviegen/moviegen/pipelines/train_pipeline.py:53]: ops.transpose -> mint.permute

File: examples/moviegen/moviegen/pipelines/train_pipeline.py [hunk 5]
```diff
@@ -42,16 +41,24 @@ def __init__(
...
+            video_emb = ops.stop_gradient(self.vae.encode(video_tokens)).astype(float32)
```
Comment (zhtmike) [inline examples/moviegen/moviegen/pipelines/train_pipeline.py:54]: better to use ms.float32

File: examples/moviegen/moviegen/schedulers/rectified_flow.py [hunk 6]
```diff
@@ -125,7 +125,9 @@ def _broadcast(self, x: Tensor) -> Tensor:
             return x
         return self.broadcast((x,))[0]
 
-    def construct(self, x: Tensor, text_embedding: Tensor, timestep: Optional[Tensor] = None) -> Tensor:
+    def construct(
```
Comment (zhtmike) [inline examples/moviegen/moviegen/schedulers/rectified_flow.py:128]: better to add docstring

File: examples/moviegen/moviegen/schedulers/rectified_flow.py [hunk 7]
```diff
@@ -139,13 +141,18 @@ def construct(self, x: Tensor, text_embedding: Tensor, timestep: Optional[Tensor
...
-            ms.float32
-        )
+        model_output = self.model(
+            x_t.to(self.model.dtype),
+            timestep,
+            ul2_emb.to(self.model.dtype),
+            metaclip_emb.to(self.model.dtype),
+            byt5_emb.to(self.model.dtype),
+        ).to(ms.float32)
+        velocity_pred = mint.chunk(model_output, 2, dim=2)[0]
```
Comment (zhtmike) [inline examples/moviegen/moviegen/schedulers/rectified_flow.py:151]: no need to do chunk here. The additional channel is used for variance prediction for iddpm only. No need for flow based model

File: examples/moviegen/tools/patch_pynative.sh [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# Patch MindSpore to add support for recompute in PyNative mode
+
+# Find the site-packages path
```
Comment (zhtmike) [inline examples/moviegen/tools/patch_pynative.sh:3]: it will easily cause bug to modify the script on fly. We may just use `if get_context("mode")==1: xxx` now

File: examples/moviegen/train.py [hunk 9]
```diff
@@ -1,299 +1,219 @@
...
-    SaveCkptCallback,
-    TimeMonitor,
-    check_cfgs_in_parser,
-    count_params,
-    init_env,
-    load_ckpt_params,
-    str2bool,
-)
-from transformers import AutoTokenizer
+from moviegen.utils import EMA, MODEL_DTYPE, MODEL_SPEC, load_ckpt_params
```
Comment (zhtmike) [inline examples/moviegen/train.py:21]: may delete callback.py as well or drop the unused callbacks

File: examples/moviegen/train.py [hunk 10]
```diff
@@ -1,299 +1,219 @@
...
-        max_rowsize=-1,
+    dataloader = create_dataloader(
+        dataset, transforms=transforms, device_num=device_num, rank_id=rank_id, **args.dataloader
     )
-    data_generator = data_generator.batch(args.batch_size, drop_remainder=True)
 
     # 5. build training utils: lr, optim, callbacks, trainer
     # 5.1 LR
-    lr = nn.WarmUpLR(learning_rate=args.start_learning_rate, warmup_steps=args.warmup_steps)
+    lr = create_scheduler(steps_per_epoch=dataloader.get_dataset_size(), **args.train.lr_scheduler)
```
Comment (zhtmike) [inline examples/moviegen/train.py:102]: I think the mindone `create_scheduler` creates a very very long array storing the learning rate (in cpu & npu), it is actually unnecessary, we can just calculate the lr on fly as `nn.WarmUpLR` does

File: examples/moviegen/configs/train/moviegen_t2i_256x256.yaml
Comment (zhtmike) [inline examples/moviegen/configs/train/moviegen_t2i_256x256.yaml:1]: nice refactor

File: examples/moviegen/moviegen/dataset/dataset.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+        if self._transforms:
+            data = self._apply_transforms(data)
+
+        return tuple(data[c] for c in self.output_columns)
+
+    def get_bucket(self, thw: Tuple[int, int, int], sample_ids: List[int]) -> Tuple[Any, ...]:
+        batch = [self._get_item(sample_id, thw) for sample_id in sample_ids]
+        return tuple(np.stack(item) for item in map(list, zip(*batch)))
+
+    def __getitem__(self, idx: int) -> Tuple[Any, ...]:
```
Comment (zhtmike) [inline examples/moviegen/moviegen/dataset/dataset.py:221]: better to have a meaning return type than `Any`

File: examples/moviegen/moviegen/pipelines/train_pipeline.py [hunk 12]
```diff
@@ -13,7 +14,6 @@ def __init__(
         network: RFlowLossWrapper,
         vae: Optional[nn.Cell] = None,
         text_encoder: Optional[nn.Cell] = None,
-        scale_factor: float = 0.13025,
         text_emb_cached: bool = True,
```
Comment (zhtmike) [inline examples/moviegen/moviegen/pipelines/train_pipeline.py:17]: better to put scale factor here, since it is independent to the VAE model (training/infernece),  only dependent on the diffusion model

File: examples/moviegen/scripts/train_t2i_256x256.sh [hunk 13]
```diff
@@ -0,0 +1,26 @@
+export ASCEND_RT_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
+# improve data loading performance for distributed training: 1
+export MS_ENABLE_NUMA=0
+# plot memory usage, feature/model: 1
+export MS_MEMORY_STATISTIC=0
+export MS_DATASET_SINK_QUEUE=4
```
Comment (zhtmike) [inline examples/moviegen/scripts/train_t2i_256x256.sh:6]: no need MS_DATASET_SINK_QUEUE since we did not support data sink yet

File: examples/moviegen/scripts/train_t2i_256x256.sh [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+output_dir=output/moviegen_t2i_256x256/$(date +"%Y.%m.%d-%H.%M.%S")
+
+msrun --bind_core=True --worker_num=8 --local_worker_num=8 --log_dir="$output_dir"  \
+python train.py \
+  --config configs/train/moviegen_t2i_256x256.yaml \
+  --env.mode 0 \
+  --env.jit_level O0 \
+  --env.max_device_memory 59GB \
+  --env.distributed=True \
```
Comment (zhtmike) [inline examples/moviegen/scripts/train_t2i_256x256.sh:19]: better to be consistent to use " " or "="
